examples/ezmsg_generator.py

```diff
@@ -0,0 +1,200 @@
+    """
+    return lambda x: reduce(lambda f, g: g.send(f), list(funcs), x)
+
+
+# State class to hold the generator instance
+class GenState(ez.State):
+    gen: Generator
+
+
+# Abstract Unit class that uses a generator for processing messages
+class Gen(ez.Unit):
+    STATE: GenState
+
+    INPUT = ez.InputStream(AxisArray)
+    OUTPUT = ez.OutputStream(AxisArray)
+
+    def initialize(self) -> None:
+        self.construct_generator()
+
+    # Method to be implemented by subclasses to construct the specific generator
+    def construct_generator(self):
+        raise NotImplementedError
+
+    # Subscriber method that sends incoming messages to the generator and publishes the result
+    @ez.subscriber(INPUT)
+    @ez.publisher(OUTPUT)
+    async def on_message(self, message: AxisArray) -> AsyncGenerator:
+        try:
+            ret = self.STATE.gen.send(message)
+            if ret is not None:
+                yield self.OUTPUT, ret
+        except (StopIteration, GeneratorExit):
+            ez.logger.debug(f"Generator closed in {self.address}")
+        except Exception:
+            ez.logger.info(traceback.format_exc())
+
+
+@consumer
+def pow(n: float) -> Generator[AxisArray, AxisArray, None]:
+    axis_arr_in = AxisArray(np.array([]), dims=[""])
+    axis_arr_out = AxisArray(np.array([]), dims=[""])
+    while True:
+        axis_arr_in = yield axis_arr_out
+        axis_arr_out = replace(axis_arr_in, data=axis_arr_in.data**n)
+
+
+class PowSettings(ez.Settings):
+    n: float
+
+
+class Pow(Gen):
+    SETTINGS: PowSettings
+
+    def construct_generator(self):
+        self.STATE.gen = pow(self.SETTINGS.n)
+
+
+@consumer
+def add(n: float) -> Generator[AxisArray, AxisArray, None]:
+    axis_arr_in = AxisArray(np.array([]), dims=[""])
+    axis_arr_out = AxisArray(np.array([]), dims=[""])
+    while True:
+        axis_arr_in = yield axis_arr_out
+        axis_arr_out = replace(axis_arr_in, data=axis_arr_in.data + n)
+
+
+class AddSettings(ez.Settings):
+    n: float
+
+
+class Add(Gen):
+    SETTINGS: AddSettings
+
+    def construct_generator(self):
+        self.STATE.gen = add(self.SETTINGS.n)
+
+
+if __name__ == "__main__":
+    print("Running offline example")
+    pipeline = compose(
+        pow(3),
+        add(12),
+        pow(0.5),
+    )
+
+    _in = AxisArray(np.arange(100), dims=["data"])
+    _out = pipeline(_in)
+    print(f"Input: {_in.data[::10]}\nOutput: {_out.data[::10]}")
+
+    print("Running online example")
+
+    class MessageSender(ez.Unit):
+        OUTPUT = ez.OutputStream(Any)
+
+        @ez.publisher(OUTPUT)
+        async def send_data(self):
+            yield self.OUTPUT, _in
+            await asyncio.sleep(1)
+            raise ez.NormalTermination
+
+    send = MessageSender()
+    pow_3 = Pow(PowSettings(3))
+    add_12 = Add(AddSettings(12))
+    pow_0_5 = Pow(PowSettings(0.5))
+    log_in = DebugLog()
+    log_out = DebugLog()
+
+    ez.run(
```

Comment:
This is the first time I have seen this usage of `run` (kwargs are units and a set of connections). All the other examples use `ez.run(SYSTEM)` where `SYSTEM` is an `ez.Collection`.

IMO it's a bit too much to introduce this new usage pattern here, in an already dense example. There should either be another very simple example that shows this `ez.run` usage or this generator example should use the more familiar ez.Collection.

Reply:
Thanks for the edit. Consider this resolved.